### ISSUE TITLE
pulumi.dynamic.Resource requires all exposed fields to be passed as `undefined`

### ISSUE THREAD

**Issue by migueloller** [pulumi/pulumi#5174]
`pulumi.dynamic.Resource` requires all exposed fields to be passed as `undefined`. For example, see the following Stripe webhook endpoint dynamic resource.

```typescript
import { ResourceProvider, Resource } from '@pulumi/pulumi/dynamic'
import { Input, Output, CustomResourceOption, secret } from '@pulumi/pulumi'
import Stripe from 'stripe'

const webhookEndpointProvider: ResourceProvider = {
  // [redacted for length] contained implementation for `create`, `diff`, `update`, and `delete`.
}

type ResourceArgs<T> = { [K in keyof T]: Input<T[K]> }
type ResourceOuts<T> = { [K in keyof T]: Output<T[K]> }

export type WebhookEndpointResourceArgs = ResourceArgs<Stripe.WebhookEndpointCreateParams>

export class WebhookEndpoint extends Resource implements ResourceOuts<Stripe.WebhookEndpoint> {
  public readonly id!: Output<Stripe.WebhookEndpoint['id']>

  // [redacted for length] contained types for output fields `object`, `api_version`, `application`, etc.

  public readonly url!: Output<Stripe.WebhookEndpoint['url']>

  constructor(name: string, args: WebhookEndpointResourceArgs, opts?: CustomResourceOptions) {
    super(
      webhookEndpointProvider,
      `stripe::WebhookEndpoint::${name}`,
      {
        id: undefined,
        object: undefined,
        api_version: undefined,
        application: undefined,
        created: undefined,
        description: undefined,
        livemode: undefined,
        metadata: undefined,
        secret: undefined,
        status: undefined,
        ...args,
      },
      opts,
    )

    this.secret = secret(this.secret)
  }
}
```

If `undefined` isn't passed for all those properties in the constructor, then they won't be exposed by the resource as outputs. Also, all of these values are shown as `output<string>`s in the preview when running `pulumi up`.

There should be a better way to indicate what fields should be exposed than passing `undefined` to the `super` call. Also, it would be nice if the type of these output fields could be explicitly defined so that a value like `livemode`, which is either `true` or `false` will appear as `output<boolean>` instead of `output<string>`.

**Comment by migueloller** [comment]
I'm still running into this issue with new dynamic resource implementations. Any plans on fixing this?

**Comment by leezen** [comment]
@migueloller Have you tried following the guidance in https://www.pulumi.com/docs/intro/concepts/resources/#the-resource-provider-interface and https://www.pulumi.com/docs/intro/concepts/resources/#dynamic-resource-outputs -- you should be able to have a strongly typed interface here

**Comment by migueloller** [comment]
@leezen, yes, I did use `ResourceProvider` as shown in the snippet, I just removed the methods to reduce the length of it.

As far as type safety of the fields, we have to manually declare each output field on the class that extends `dynamic.Resource` and that's consistent with what's documented in https://www.pulumi.com/docs/intro/concepts/resources/#dynamic-resource-outputs

> If you need to access the outputs of your custom resource outside it with strong typing support, declare each output property returned in the outs property by your create function as a class member of the pulumi.dynamic.Resource itself.

The main issue here isn't the type safety (although it would be nice to have those fields be type safe automatically), but instead the Pulumi runtime not exposing outputs unless they are passed in when calling `super` on the class that extends `dynamic.Resource`. If I try to access an output like so: `resource.output`, unless I pass `undefined` or some other value in the `super` call then the value simply won't be there.

Here's an example:

```
const we = new WebhookEndpoint('we', { url })

export const webhookEndpointSecret = we.secret
```

Unless `super` is called with `super({ secret: undefined })` in the dynamic resource definition, because `secret` is not actually an argument for the resource, then the output itself will be undefined and attempting to do anything with `we.secret` would fail.

**Comment by lukehoban** [comment]
Yes - it is currently by design that `undefined` must be passed as inputs for outputs, as noted in https://www.pulumi.com/docs/intro/concepts/resources/#dynamic-resource-outputs.  This design is certainly "awkward" though, and it's worth tracking this as part of potential future re-design of Dynamic Providers.
